proposals/Fleetctl-trigger.md

```diff
@@ -0,0 +1,95 @@
+
+### New endpoint `GET /trigger?name={:name}` 
+- The request handler first calls `ds.Lock` to check if the named schedule is locked.
+- If the named schedule is unlocked, request handler sends a trigger signal on the schedule's
+  trigger channel and the server responds with status `202 Accepted`.  
+- If the named schedule is locked (presumably because the schedule is currently running), the server
+  responds with status `409 Conflict` and includes a message indicating the schedule is currently
+  locked. It is up to the user to retry. To facilitate retries, the response message could be expanded to
+  include additional status information, such as the expiration time of the current lock.
+
+### Schedule locks
+- Currently, lock duration is based on the schedule interval. 
+  - Once an instance takes the lock, it will hold the lock for the duration of the interval, even
+    after it has completed the jobs in the schedule. 
+  - For long-running jobs, the lock may expire before the current instance completes its run,
+    meaning that it is currently possible for another instance to start an overlapping job.
+  - If the lockholder instance is terminated or killed, locks are not released, which may frustrate
+    a user's attempt to configure a shorter schedule interval before the lock held by the dead
+    instance expires.
+
+- Under this proposal, locks become more dynamic.
+  - Current lockholder releases its lock once it finishes running the schedule.
+  - Graceful shutdown process handles release of locks upon termination signals. Jobs are
+    preemptable and an interrupt function must be specified for each job, 
+    e.g., `schedule.New(...).WithJob("job_name", jobFn, interruptFn)`.
+  - As a fallback for cases that can't be handled via graceful shutdown (e.g., `SIGKILL`), the
+    expiration for a lock is initially set to a relatively short default duration (e.g., 5 minutes).
+    The expiration is then periodically extended by the current instance so long as scheduled jobs
+    are running.  If the current instance dies without graceful shutdown, the lock will only be held
+    by the dead instance for a short period.
+
+### Additional UX considerations 
+- What are some potential options that would be useful for the `fleetctl trigger` command?
+  - Request the current status of the the named schedule without triggering a new run.
+    For example, `--status` could provide the user with the running time of the schedule (this would
+    require that we expand the `locks` table to include additional timestamp information, such as
+    lock start time and lock release time).
+  - Other useful options?
+
+- What rules should determine when the interval ticker resets? Consider the following cases where
+  `s.scheduleInterval = 1*time.Hour`: 
+  - The schedule is triggered at 55 minutes into the 1-hour interval and takes 1 minute to complete.
+    When should the schedule run again? 
+    (a) after 4 minutes; 
+    (b) after 1 hour; 
```

Comment:
I would expect it to run 1 hour after the last time it was run (start time) ie option b.

Reply:
How would you handle the second case?

Reply:
@lucasmrod, so you are saying the answer to each of these [questions](https://github.com/fleetdm/fleet/pull/8127#discussion_r991534006) should be "yes"?

Reply:
Yes :)

Reply:
> How would you handle a case where the job takes longer than the interval?
Probably just skip the next run. It would fail to get the lock and have to wait until the next tick.